### PR TITLE
http: fix metadata bug with local replies

### PR DESCRIPTION
**Pull request by paul-r-gall**
Commit Message: http: fix metadata bug with local replies
Additional Description: When Filter1 returns StopIteration from decodeHeaders ContinueAll from decodeMetadata, and Filter2 does sendLocalReply from decodeHeaders, it triggers an ASSERT in decodeMetadata, because metadata continues to be decoded after sendLocalReply from headers. This PR fixes that bug.
Risk Level: low
Testing: integration testing
Docs Changes: none
Release Notes: none

### CODE REVIEW

**Review by alyssawilk — COMMENTED** [review]
code and test LGTM but I'd be inclined to runtime guard in the unlikely event the fix is more problematic than !fix.  thoughts?
/wait-any
/retest

**Comment by paul-r-gall** [comment]
Sure; does default-on sound reasonable?

**Review by alyssawilk — APPROVED** [review]

File: changelogs/current.yaml [hunk 1]
```diff
@@ -160,6 +160,10 @@ bug_fixes:
 - area: redis
   change: |
     Fixed a bug where redis key with % in the key is failing with a validation error.
+- area: http
+  change: |
+    Fixed a bug that could cause an metadata to be decoded after a local reply has been triggered.
```
Comment (alyssawilk) [inline changelogs/current.yaml:165]: I'd say cause metadata rather than an metadata but enh?

**Comment by alyssawilk** [comment]
oops needs main merge.  adding in @nezdolik for !google OK on this

File: test/integration/filters/local_reply_during_decoding_filter.cc [hunk 2]
```diff
@@ -20,6 +20,16 @@ class LocalReplyDuringDecode : public Http::PassThroughFilter {
                                        "");
     return Http::FilterHeadersStatus::StopIteration;
   }
+
+  Http::FilterDataStatus decodeData(Buffer::Instance&, bool) override {
+    ASSERT(false);
```
Comment (nezdolik) [inline test/integration/filters/local_reply_during_decoding_filter.cc:25]: nit:add explanatory comment that this method should never be invoked in tests, same for one below

**Comment by nezdolik** [comment]
Looks good to me, thanks! (modulo minor nit)

**Review by alyssawilk — APPROVED** [review]

**Pull request merged**